File: problems/Swift/11.盛最多水的容器.swift
/**
 * 11.盛最多水的容器 [Medium]
 * Time Complexity: O(n), Space Complexity: O(1)
 * https://leetcode-cn.com/problems/container-with-most-water/
 */

class Solution {
  func maxArea(_ height: [Int]) -> Int {
    var left = 0, right = height.count - 1, maxArea = 0
    while left < right {
      let area = (right - left) * min(height[left], height[right])
      maxArea = max(maxArea, area)
      if height[left] < height[right] {
        left += 1
      } else {
        right -= 1
      }
    }
    return maxArea
  }
}

File: problems/rust-solutions/src/11.盛最多水的容器.rs
// 11.盛最多水的容器 [Medium]
// Time Complexity: O(n), Space Complexity: O(1)
// https://leetcode-cn.com/problems/container-with-most-water/

impl Solution {
  pub fn max_area(height: Vec<i32>) -> i32 {
    let mut max_area = 0;
    let mut left = 0;
    let mut right = height.len() - 1;

    while left < right {
      let area = (right - left) as i32 * (height[left].min(height[right]));
      max_area = max_area.max(area);

      if height[left] < height[right] {
        left += 1;
      } else {
        right -= 1;
      }
    }

    return max_area;
  }
}
